### PR TITLE
Add payment channels support

### PR DESCRIPTION
**Pull request by ghost**
This PR adds support for the Zondax wasm payment channels library.  

# Why

The purpose of the retrieval browser extension is to create a _marketplace_ for CIDs.  However, prior to this PR, all transfers from browser-to-browser were free because there was no way to pay to provider of the content.  After this PR, all transfers actually result in payment to the provider, with multiple on-chain interactions to ensure payment will be made upon completion of the transfer.  Thus, we have a marketplace where buyers pay sellers using Filecoin.

# What it does

This PR incorporates all of the Filecoin Payment Channel operations in browser-to-browser transfers:  

 - create pch
 - create signed vouchers
 - check signature on signed vouchers
 - update the payment channel with highest value signed voucher
 - settle the payment channel (begins 12-hour waiting period to collect)
 - collect the payment channel

# Build Notes

Currently, this PR requires a custom version of the filecoin-signing-tools library.  You need to build it from this PR:  https://github.com/Zondax/filecoin-signing-tools/pull/317.

The wasm compilation requires installing the rust toolchain (`curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`, use the default options, now you have rust).  Type `make` in the base directory `filecoin-signing-tools` and it will build the wasm binaries.

After compiling, cd into `filecoin-singing-tools/signer-npm/pkg` and type `yarn link`.  Then cd into `<your extension directory>/src/background/lotus-client/` and type `yarn link "@zondax/filecoin-signing-tools"`.  Then `yarn && yarn build` to rebuild the extension.

### CODE REVIEW

**Review by aistoc — APPROVED** [review]

File: src/background/lotus-client/Lotus.js [hunk 1]
```diff
@@ -47,131 +63,381 @@ class Lotus {
...
+
+    ports.postLog(`DEBUG: Lotus.createPaymentChannel: [from:${fromAddr}, fromKey:*******************, to:${toAddr}, amount:${amountAttoFil}]`);
+
+    let nonce = await this.getNonce(fromAddr);
+    ports.postLog(`DEBUG: Lotus.createPaymentChannel: nonce=${nonce}`);
+
+    //
+    // Generate the PCH create message
+    //
+    var signedCreateMessage;
```
Comment (ionutiorgulescu) [inline src/background/lotus-client/Lotus.js]: Don't need this.

File: src/background/lotus-client/Lotus.js [hunk 2]
```diff
@@ -47,131 +63,381 @@ class Lotus {
...
+   */
+  async updatePaymentChannel(pch, signedVoucher) {
+    const toAddr = this.wallet;
+    const toPrivateKeyBase64 = this.privateKeyBase64;
+    ports.postLog(`DEBUG: Lotus.updatePaymentChannel:\n  pch=${pch}\n  toAddr=${toAddr}\n  toPrivateKeyBase64=*****************\n  signedVoucher=${signedVoucher}`);
+    
+    //
+    // Generate update PCH message
+    //
+    var signedUpdateMessage;
```
Comment (ionutiorgulescu) [inline src/background/lotus-client/Lotus.js]: same here don't need this init

File: src/background/lotus-client/Lotus.js [hunk 3]
```diff
@@ -47,131 +63,381 @@ class Lotus {
...
+    try {
+      let nonce = await this.getNonce(toAddr);
+      let update_paych_message = signer.updatePymtChanWithFee(pch, toAddr, signedVoucher, nonce, "10000000", "16251176117", "140625002") // gas limit, fee cap, premium
```
Comment (ionutiorgulescu) [inline src/background/lotus-client/Lotus.js]: why not use camelCase?

File: src/background/lotus-client/Lotus.js [hunk 4]
```diff
@@ -47,131 +63,381 @@ class Lotus {
...
+    const toPrivateKeyBase64 = this.privateKeyBase64;
+    ports.postLog(`DEBUG: Lotus.settlePaymentChannel:\n  pch=${pch}\n  toAddr=${toAddr}\n  toPrivateKeyBase64=**************`);
+    
+    //
+    // Generate Settle PCH message
+    //
+    var signedSettleMessage;
+    try {
+      let nonce = await this.getNonce(toAddr);
+      let settle_paych_message = signer.settlePymtChanWithFee(pch, toAddr, nonce, "10000000", "16251176117", "140625002") // gas limit, fee cap, premium)
```
Comment (ionutiorgulescu) [inline src/background/lotus-client/Lotus.js]: camelCase?

File: src/background/lotus-client/Lotus.js [hunk 5]
```diff
@@ -47,131 +63,381 @@ class Lotus {
...
+   */
+  async collectPaymentChannel(pch) {
+    const toAddr = this.wallet;
+    const toPrivateKeyBase64 = this.privateKeyBase64;
+    ports.postLog(`DEBUG: Lotus.collectPaymentChannel:\n  pch=${pch}\n  toAddr=${toAddr}\n  toPrivateKeyBase64=*************`);
+    
+    //
+    // Generate Collect PCH message
+    //
+    var signedCollectMessage;
```
Comment (ionutiorgulescu) [inline src/background/lotus-client/Lotus.js]: again don't need it

File: src/background/retrieval-market/Provider.js [hunk 6]
```diff
@@ -115,9 +126,21 @@ class Provider {
     });
   };
 
-  async handleNewDeal({ dealId, cid, params }, sink) {
-    ports.postLog("DEBUG: Provider.handleNewDeal()")
-    ports.postLog(`DEBUG: handling new deal ${dealId}`);
+  setOrVerifyPaymentChannel(message) {
```
Comment (ionutiorgulescu) [inline src/background/retrieval-market/Provider.js:135]: why not to split in two functions to be more clear?

File: src/background/lotus-client/Lotus.js (hunk 1 above)
Comment (ghost) [reply src/background/lotus-client/Lotus.js]: Not a JS expert, so I honestly don't know, but if I declare it inside the `try {...}` block, won't it go out of scope at the end of the `try{}catch{}`?  I need to use that value later, after the try catch.

For now marked this line as a `// TODO:  does this need to be declared out here?`

File: src/background/lotus-client/Lotus.js (hunk 3 above)
Comment (mgoelzer) [reply src/background/lotus-client/Lotus.js:283]: Marked as a TODO

File: src/background/lotus-client/Lotus.js (hunk 2 above)
Comment (ghost) [reply src/background/lotus-client/Lotus.js]: Marked as todo also.

File: src/background/retrieval-market/Provider.js (hunk 6 above)
Comment (ghost) [reply src/background/retrieval-market/Provider.js:135]: Yeah, this is probably a good point.  I logged them as TODO's now so I can merge this code without changes that may break it (it's rather fragile :-)), but I would like to fix this eventually.  Also marked the code duplication between `dealStatuses.paymentSent` and `dealStatuses.lastPaymentSent`.

File: src/background/lotus-client/Lotus.js (hunk 3 above)
Comment (ghost) [reply src/background/lotus-client/Lotus.js]: Fixed

File: src/background/lotus-client/Lotus.js (hunk 4 above)
Comment (ghost) [reply src/background/lotus-client/Lotus.js]: Fixed

**Pull request merged**